torchvision/prototype/datasets/_builtin/coco.py

```diff
@@ -26,24 +28,144 @@
+
+
+# TODO: this is in desperate need of comments. They will be added as soon as we are sure this is the way to go
+class CocoAnnotationsCollater(IterDataPipe[Tuple[Dict[str, Any], Dict[str, Dict[str, Any]]]]):
+    def __init__(self, partial_ann_dps: Dict[str, IterDataPipe[Tuple[List[Dict[str, Any]], Dict[str, Any]]]]) -> None:
+        self.partial_ann_dps = partial_ann_dps
+
+    def __iter__(self) -> Iterator[Tuple[Dict[str, Any], Dict[str, Dict[str, Any]]]]:
+        exhausted_ann_types = set(Coco._ANN_TYPES) - self.partial_ann_dps.keys()
+
+        if len(self.partial_ann_dps) == 1:
+            ann_type, dp = next(iter(self.partial_ann_dps.items()))
+            for ann, image_meta in dp:
+                anns = {ann_type: ann}
+                for exhausted_ann_type in exhausted_ann_types:
+                    anns[exhausted_ann_type] = []
+                yield image_meta, anns
+            return
+
+        dps = [(ann_type, iter(dp), dict()) for ann_type, dp in self.partial_ann_dps.items()]
+        while dps:
+            ref_ann_type, ref_dp, ref_buffer = dps[0]
+            try:
+                ref_anns, ref_image_meta = next(ref_dp)
+            except StopIteration:
+                if not ref_buffer:
+                    exhausted_ann_types.add(ref_ann_type)
+                    del dps[0]
+                    continue
+
+                ref_anns, ref_image_meta = ref_buffer.pop(next(iter(ref_buffer.keys())))
+
+            anns = {ref_ann_type: ref_anns}
+            ref_key = ref_image_meta["id"]
+
+            for idx, (other_ann_type, other_dp, buffer) in enumerate(dps[1:], 1):
+                if ref_key in buffer:
+                    other_anns, _ = buffer.pop(ref_key)
+                    anns[other_ann_type] = other_anns
+                    continue
+
+                for other_anns, other_image_meta in other_dp:
+                    other_key = other_image_meta["id"]
+                    if other_key == ref_key:
+                        anns[other_ann_type] = other_anns
+                        break
+
+                    buffer[other_key] = other_anns, other_image_meta
+                else:
+                    if buffer:
+                        anns[other_ann_type] = []
+                    else:
+                        exhausted_ann_types.add(other_ann_type)
+                        del dps[idx]
+
+            for exhausted_ann_type in exhausted_ann_types:
+                anns[exhausted_ann_type] = []
+
+            yield ref_image_meta, anns
 
 
 class Coco(Dataset):
+    def _decode_instances_anns(self, anns: List[Dict[str, Any]], image_meta: Dict[str, Any]) -> Dict[str, Any]:
+        labels = [ann["category_id"] for ann in anns]
+        categories = [self.info.categories[label] for label in labels]
+        return dict(
+            areas=torch.tensor([ann["area"] for ann in anns]),
+            iscrowd=torch.tensor([ann["iscrowd"] for ann in anns], dtype=torch.bool),
```

Comment:
FYI COCO also contains segmentation masks for each instance (and sometimes even keypoints depending on the instantiation).

We will need to decide at that point what will be the format representation for the segmentation masks (they are stored as RLE polygons sometimes).

Reply:
Yes, I'm aware of that, and I'll support for that in a follow-up PR starting with segmentations. This improvement makes it really easy to load additional annotations. For example `datasets.load("coco", instances=True, captions=True)` will load both and `keypoints` will be added to that list.